text/3773-must-use-output.md

```diff
@@ -0,0 +1,124 @@
+
+As an example, this code:
+
+```rust
+fn main() {
+    let mut vec = Vec::new();
+    vec.push(42);
+}
+```
+
+will emit a warning saying "`vec` is not used after the call to `push`" because `Vec::push` has `self` marked with this attribute:
+
+```rust
+impl<T> Vec<T> {
+    pub fn push(#[must_use_output] &mut self, item: T) { /* ... */ }
+}
+```
+
+This is conceptually similar to `#[must_use]` on the function, but for an output parameter rather than the return value. Like `#[must_use]`, it helps you (or downstream consumers) to catch likely mistakes.
+
+A common case when this is useful is when the function has no side effects other than mutation through the passed reference and allocation. A typical example is methods modifying collections, builders or similar objects.
+
+Note that this is currently subject to some limitations: if the reference was passed in as an argument, returned from another function or obtained from a pointer the warning will not trigger. This may change in the future and it may start producing more warnings in some cases.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Whenever a function parameter is marked with `#[must_use_output]`, and the caller does not later access the value, a warning is emitted similar to using the `+=` operator on integers and later not using the value.
+
+It is an error to put this attribute on parameters with types other than reference types. Raw pointers, smart pointers, and generics that hide the reference, are forbidden, mainly for simplicity of initial implementation. They can be implemented and allowed later, if it's even possible at all.
+
+The compiler makes no distinction whether the reference is unique or shared because writes can happen through both; there doesn't seem to be a reason to forbid `Freeze` references and there is at least one case when even a `Freeze` reference is useful.
+
+The standard library functions get annotated with this attribute as well, including but not limited to:
+
+- Modifications of collections that don't return a value
+- `core::mem::swap`
+- `Clone::clone`
+- `*_assign` methods in `core::ops` traits
+- `OpenOptions` and similar builders
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Adds another attribute to the language, complicating it
+- In some cases analysis will fail and perhaps people will have false sene of security
+- People could misuse it to attempt to enforce soundness of `unsafe` code (which it cannot do)
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- We could also simply not do this but the potential mistakes it catches are real.
+- The name could be something different and preferably shorter. The name used here was suggested by Josh Triplett and is pretty readable and it can be used before stabilization in the absence of better one. A pretty short and clear `#[must_read]` was also suggested by Kevin Reid.
```

Comment:
For the record, while I *mildly* prefer `must_use_output`, I would happily check a box for `must_read`, and it seems like a fine choice too.

Reply:
maybe `#[caller_must_use]`?

Reply:
@kennytm Sadly still pretty long but I like it better than `must_use_output`. It at least doesn't pose "what do you mean by the output?" question.

Reply:
@Kixunil Agreed. `must_use` and `must_read` don't answer "who". `must_use_output` specifies that indirectly: "the one that gets this as an output must use it". But `caller_must_use` specifies it *directly*. That seems pretty clear to me.